

#Menentukan Objektif

Tabel yang akan digunakan pada project adalah 3 tabel yang saling berelasi yaitu:
1.   olist_geolocation_dataset
2.   olist_order_items_dataset
3.   olist_sellers_dataset

fokus analisa yang dilakukan pada project ini yaitu:
1.   Banyaknya jumlah order item perbulan berdasarkan seller
2.   Banyaknya jumlah order item perbulan berdasarkan kota
1.   Persebaran seller berdasarkan kota




















# Open Data

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from scipy import stats
from datetime import datetime

In [ ]:
data = pd.read_csv("drive/MyDrive/Colab Notebooks/Sales.csv")
data.head(10)

zip_code        lat       lang                         seller_id  \
0      2518 -23.511112 -46.666213  830379336fad8c6f3b15a4a3ddb5c66e   
1      2518 -23.511112 -46.666213  830379336fad8c6f3b15a4a3ddb5c66e   
2      2518 -23.511112 -46.666213  830379336fad8c6f3b15a4a3ddb5c66e   
3      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
4      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
5      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
6      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
7      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
8      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   
9      2537 -23.492697 -46.654851  bf84056e679dbe9c69929847a40e338f   

   city_name code_state                        product_id  order_item  price  \
0  sao paulo         SP  481587fdec29c4d63a801769b4e5c578         2.0  67.00   
1  sao paulo         SP  481587fdec29c4d63a801769b4e5c578         1.0  67.00   
2  sao paulo         SP  b7d8104067b422204524dcd4706894b1         2.0  67.00   
3  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  47.61   
4  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  46.33   
5  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  53.69   
6  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  45.36   
7  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  59.89   
8  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         1.0  50.21   
9  sao paulo         SP  437c05a395e9e47f9762e677a7068ce7         2.0  46.33   

   shipping_limit_date  
0  2018-04-26 18:32:42  
1  2018-04-26 18:32:42  
2  2018-07-30 11:33:09  
3  2018-02-28 12:48:36  
4  2018-02-27 14:30:38  
5  2018-04-23 12:30:17  
6  2018-02-27 03:31:22  
7  2018-03-23 17:09:46  
8  2018-03-05 03:55:52  
9  2018-02-28 02:10:26

# EDA

Data specialist menggunakan EDA untuk membedakan kumpulan data apa yang dapat diungkapkan lebih jauh di luar pemodelan data formal atau tugas pengujian hipotesis. Hal ini memungkinkan mereka untuk mendapatkan temuan mendalam tentang variabel dalam kumpulan data dan hubungannya.

EDA dapat membantu mendeteksi kesalahan, mengidentifikasi outlier dalam kumpulan data, memahami hubungan antar data, menggali faktor-faktor penting, menemukan pola dalam data, dan memberikan wawasan baru. EDA sangat bermanfaat untuk analisis statistik. Data profesional sering kali menggunakan EDA untuk menemukan hasil yang valid dan relevan sebagai arah tujuan bisnis yang diinginkan.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500000 entries, 0 to 499999
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   zip_code             500000 non-null  int64  
 1   lat                  500000 non-null  float64
 2   lang                 500000 non-null  float64
 3   seller_id            455985 non-null  object 
 4   city_name            455985 non-null  object 
 5   code_state           455985 non-null  object 
 6   product_id           455985 non-null  object 
 7   order_item           455985 non-null  float64
 8   price                455985 non-null  float64
 9   shipping_limit_date  455985 non-null  object 
dtypes: float64(4), int64(1), object(5)
memory usage: 38.1+ MB


info() digunakan untuk melihat informasi dari tabel. berdasarkan data info jumlah data pada project ini adalah 500000 data dengan jumlah kolom 10. tipe data integer berjumlah 1, float berjumlah 4, dan object berjumlah 5. Pada data info juga kita bisa mengetahui jumlah data yang tidak null berdasarkan kolom.

In [ ]:
data.isnull().sum()

zip_code                   0
lat                        0
lang                       0
seller_id              44015
city_name              44015
code_state             44015
product_id             44015
order_item             44015
price                  44015
shipping_limit_date    44015
dtype: int64

Data null merupakan data kosong/missing values untuk mengetahui Jumlah pasti data null pada tiap kolom menggunakan isnull().sum()

In [ ]:
data['shipping_limit_date'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 500000 entries, 0 to 499999
Series name: shipping_limit_date
Non-Null Count   Dtype 
--------------   ----- 
455985 non-null  object
dtypes: object(1)
memory usage: 3.8+ MB


terdapat ketidak konsistenan format tipe data berdasarkan hasil data.info() dimana tipe data shipping_limit_date masih berupa object dan bukan datetime.

In [ ]:
data2 = data.duplicated(subset=['product_id','shipping_limit_date'], keep ='first')
print(data2)
data.iloc[[0, 1]]

0         False
1          True
2         False
3         False
4         False
          ...  
499995     True
499996     True
499997     True
499998     True
499999     True
Length: 500000, dtype: bool


zip_code        lat       lang                         seller_id  \
0      2518 -23.511112 -46.666213  830379336fad8c6f3b15a4a3ddb5c66e   
1      2518 -23.511112 -46.666213  830379336fad8c6f3b15a4a3ddb5c66e   

   city_name code_state                        product_id  order_item  price  \
0  sao paulo         SP  481587fdec29c4d63a801769b4e5c578         2.0   67.0   
1  sao paulo         SP  481587fdec29c4d63a801769b4e5c578         1.0   67.0   

   shipping_limit_date  
0  2018-04-26 18:32:42  
1  2018-04-26 18:32:42

mengecek apakah ada data duplikat atau tidak dengan menggunakan duplicate(subset=[nama kolom], keep ='first'). Artinya nanti data baris pertama akan di identifikasi menjadi unik sedangkan data baris selanjutnya yang memiliki value sama akan di identifikasi duplikat.
contohnya pada kasus diatas data dari baris ke 0 dan 1 terdapat duplikat berdasarkan product_id dan shipping_limit_date yang sama.

In [ ]:
len_city = data['city_name'].unique()
len_city

array(['sao paulo', nan, 'sao paulo sp', 'uberaba', 'itau de minas',
       'passos', 'araxa', 'sp / sp', 'sao pauo', 'fronteira',
       'rio de janeiro', 'rio de janeiro \\rio de janeiro'], dtype=object)

terdapat ketidak konsistenan nama kota. untuk mencari nama kota tanpa duplikat yang berada pada data menggunakan perintah unique()

# Data Cleaning

In [ ]:
data = data.dropna(subset=["seller_id", "city_name", "code_state", "product_id", "order_item", "price", "shipping_limit_date"])
data.isnull().sum()

zip_code               0
lat                    0
lang                   0
seller_id              0
city_name              0
code_state             0
product_id             0
order_item             0
price                  0
shipping_limit_date    0
dtype: int64

Handling Missing Values atau Data Nan menggunakan dropna(subset=kolom).

In [ ]:
data['shipping_limit_date'] = pd.to_datetime(data["shipping_limit_date"])
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 455985 entries, 0 to 499999
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   zip_code             455985 non-null  int64         
 1   lat                  455985 non-null  float64       
 2   lang                 455985 non-null  float64       
 3   seller_id            455985 non-null  object        
 4   city_name            455985 non-null  object        
 5   code_state           455985 non-null  object        
 6   product_id           455985 non-null  object        
 7   order_item           455985 non-null  float64       
 8   price                455985 non-null  float64       
 9   shipping_limit_date  455985 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(4), int64(1), object(4)
memory usage: 38.3+ MB


Mengubah tipe data shiping_limit_date dari object menjadi datetime menggunakan perintah pd.to_datetime(). tipe data shiping_limit_date telah berganti menjadi datetime berdasarkan hasil informasi data.

In [ ]:
data = data.drop_duplicates(subset=['product_id', 'shipping_limit_date'], keep='last')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10169 entries, 17223 to 499999
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   zip_code             10169 non-null  int64         
 1   lat                  10169 non-null  float64       
 2   lang                 10169 non-null  float64       
 3   seller_id            10169 non-null  object        
 4   city_name            10169 non-null  object        
 5   code_state           10169 non-null  object        
 6   product_id           10169 non-null  object        
 7   order_item           10169 non-null  float64       
 8   price                10169 non-null  float64       
 9   shipping_limit_date  10169 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(4), int64(1), object(4)
memory usage: 873.9+ KB


menghapus data duplikat berdasarkan kolom referensi dan mempertahankan data terakhir sebagai data unik. menggunakan perintah drop_duplicate(subset=namakolom, keep='last')

In [ ]:
data = data.replace(['sao paulo sp','sao pauo', 'sp / sp','rio de janeiro \\rio de janeiro' ],
 ['sao paulo','sao paulo','sao paulo','rio de janeiro'])

In [ ]:
len_city1 = data['city_name'].unique()
len_city1

array(['sao paulo', 'itau de minas', 'passos', 'uberaba', 'araxa',
       'fronteira', 'rio de janeiro'], dtype=object)

untuk memperbaiki ketidak konsistenan data menjadi data yang sesuai dengan value yang semestinya menggunakan perintah replace()

# Data Manipulation

## **Parsing Data Berdasarkan Tahun**

In [ ]:
data = data.set_index('shipping_limit_date')
data1 = data.loc['2016-01-01 00:00:00':'2016-12-31 23:59:59']
data1.reset_index(inplace=True)
data2 = data.loc['2017-01-01 00:00:00':'2017-12-31 23:59:59']
data2.reset_index(inplace=True)
data3 = data.loc['2018-01-01 00:00:00':'2018-12-31 23:59:59']
data3.reset_index(inplace=True)
data4 = data.loc['2019-01-01 00:00:00':'2019-12-31 23:59:59']
data4.reset_index(inplace=True)
data5 = data.loc['2020-01-01 00:00:00':'2020-12-31 23:59:59']
data5.reset_index(inplace=True)

<ipython-input-13-9a48fcf681a5>:2: FutureWarning: Value based partial slicing on non-monotonic DatetimeIndexes with non-existing keys is deprecated and will raise a KeyError in a future Version.
  data1 = data.loc['2016-01-01 00:00:00':'2016-12-31 23:59:59']
<ipython-input-13-9a48fcf681a5>:4: FutureWarning: Value based partial slicing on non-monotonic DatetimeIndexes with non-existing keys is deprecated and will raise a KeyError in a future Version.
  data2 = data.loc['2017-01-01 00:00:00':'2017-12-31 23:59:59']
<ipython-input-13-9a48fcf681a5>:6: FutureWarning: Value based partial slicing on non-monotonic DatetimeIndexes with non-existing keys is deprecated and will raise a KeyError in a future Version.
  data3 = data.loc['2018-01-01 00:00:00':'2018-12-31 23:59:59']
<ipython-input-13-9a48fcf681a5>:8: FutureWarning: Value based partial slicing on non-monotonic DatetimeIndexes with non-existing keys is deprecated and will raise a KeyError in a future Version.
  data4 = data.loc['2019-01-

jadikan data dengan format date menjadi index lalu cari rentang waktu berdasarkan tahun menggunakan perintah loc[:]. lalu kembalikan index menjadi kolom dengan menggunakan perintah reset_index()

## **Banyaknya jumlah order item perbulan berdasarkan seller tahun 2016**

In [ ]:
data6 = pd.DataFrame(data1[['shipping_limit_date', 'seller_id', 'order_item']])
data6['month'] = data6['shipping_limit_date'].dt.month
data6['year'] = data6['shipping_limit_date'].dt.year
data6 = data6.drop(['shipping_limit_date'], axis=1)
grouped = data6.groupby(['month', 'year', 'seller_id'])['order_item'].aggregate(['count']).reset_index()
grouped['order_item'] = grouped['count']
grouped = grouped.drop(['count'], axis=1)
grouped.head(2)

month  year                         seller_id  order_item
0     10  2016  024b564ae893ce8e9bfa02c10a401ece           1
1     10  2016  46dc3b2cc0980fb8ec44634e21d2718e          10

##**Banyaknya jumlah order item perbulan berdasarkan seller tahun 2017**

In [ ]:
data7 = pd.DataFrame(data2[['shipping_limit_date', 'seller_id', 'order_item']])
data7['month'] = data7['shipping_limit_date'].dt.month
data7['year'] = data7['shipping_limit_date'].dt.year
data7 = data7.drop(['shipping_limit_date'], axis=1)
grouped1 = data7.groupby(['month', 'year','seller_id'])['order_item'].aggregate(['count']).reset_index()
grouped1['order_item'] = grouped1['count']
grouped1 = grouped1.drop(['count'], axis=1)
grouped1.head(2)

month  year                         seller_id  order_item
0      1  2017  46dc3b2cc0980fb8ec44634e21d2718e          13
1      1  2017  5f0057b677eb963672bf05dfa3f16c34           1

##**Banyaknya jumlah order item perbulan berdasarkan seller tahun 2018**

In [ ]:
data8 = pd.DataFrame(data3[['shipping_limit_date', 'seller_id', 'order_item']])
data8['month'] = data8['shipping_limit_date'].dt.month
data8['year'] = data8['shipping_limit_date'].dt.year
data8 = data8.drop(['shipping_limit_date'], axis=1)
grouped2 = data8.groupby(['month', 'year','seller_id'])['order_item'].aggregate(['count']).reset_index()
grouped2['order_item'] = grouped2['count']
grouped2 = grouped2.drop(['count'], axis=1)
grouped2.head(2)

month  year                         seller_id  order_item
0      1  2018  00ee68308b45bc5e2660cd833c3f81cc          39
1      1  2018  04e38a233f726ef161074ab1c9b0399a           2

In [ ]:
concatsel = pd.concat([grouped,grouped1, grouped2])
concatsel['month'] = concatsel['month'].astype(str)
concatsel['month'] = concatsel['month'].replace(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
 ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August','September', 'October', 'November','December'])
concatsel

month  year                         seller_id  order_item
0      October  2016  024b564ae893ce8e9bfa02c10a401ece           1
1      October  2016  46dc3b2cc0980fb8ec44634e21d2718e          10
2      October  2016  5f0057b677eb963672bf05dfa3f16c34           2
3      October  2016  6cf476a4ca74498db55cbccdaa9dcfb6           1
4      October  2016  76d5af76d0271110f9af36c92573f765           3
..         ...   ...                               ...         ...
835  September  2018  8bd0f31cf0a614c658f6763bd02dea69           1
836  September  2018  98dac6635aee4995d501a3972e047414           1
837  September  2018  99a54764c341d5dc80b4a8fac4eba3fb           1
838  September  2018  b561927807645834b59ef0d16ba55a24           4
839  September  2018  bf00385a5f7fc1ef39a13c2e9ee50a5f           1

[1514 rows x 4 columns]

##**Banyaknya jumlah order item perbulan berdasarkan kota tahun 2016**

In [ ]:
city = pd.DataFrame(data1[['shipping_limit_date', 'city_name', 'order_item']])
city['month'] = city['shipping_limit_date'].dt.month
city['year'] = city['shipping_limit_date'].dt.year
city = city.drop(['shipping_limit_date'], axis=1)
th2016 = city.groupby(['month','year','city_name'])['order_item'].aggregate(['count']).reset_index()
th2016['order_item'] = th2016['count']
th2016 = th2016.drop(['count'], axis=1)
th2016.head(2)

month  year       city_name  order_item
0     10  2016  rio de janeiro          10
1     10  2016       sao paulo          31

##**Banyaknya jumlah order item perbulan berdasarkan kota tahun 2017**

In [ ]:
city1 = pd.DataFrame(data2[['shipping_limit_date', 'city_name', 'order_item']])
city1['month'] = city1['shipping_limit_date'].dt.month
city1['year'] = city1['shipping_limit_date'].dt.year
city1 = city1.drop(['shipping_limit_date'], axis=1)
th2017 = city1.groupby(['month', 'year', 'city_name'])['order_item'].aggregate(['count']).reset_index()
th2017['order_item'] = th2017['count']
th2017 = th2017.drop(['count'], axis=1)
th2017.head(2)

month  year       city_name  order_item
0      1  2017  rio de janeiro          13
1      1  2017       sao paulo          65

##**Banyaknya jumlah order item perbulan berdasarkan kota tahun 2018**

In [ ]:
city2 = pd.DataFrame(data3[['shipping_limit_date', 'city_name', 'order_item']])
city2['month'] = city2['shipping_limit_date'].dt.month
city2['year'] = city2['shipping_limit_date'].dt.year
city2 = city2.drop(['shipping_limit_date'], axis=1)
th2018 = city2.groupby(['month', 'year','city_name'])['order_item'].aggregate(['count']).reset_index()
th2018['order_item'] = th2018['count']
th2018 = th2018.drop(['count'], axis=1)
th2018.head(2)

month  year       city_name  order_item
0      1  2018          passos           1
1      1  2018  rio de janeiro          17

Melalukan Grouping dan Agregat pada kolom yang diinginkan. pertama buat variabel baru pada kode variabel barunya adalah city2. setelah itu ektrak bulan dan tahun menggunakan perintah dt.month dan dt.year. dan masukan ke kolom baru bernama month dan year. hapus kolom shipping_limit_date karena sudah tidak dibutuhkan. lakukan grouping dan agregat untuk mendapatkan jumlah order item perbulan berdasarkan seller. ubah nama kolom count menjadi order_item. lalu hapus kolom count.

In [ ]:
concatcity = pd.concat([th2016, th2017, th2018])
concatcity['month'] = concatcity['month'].astype(str)
concatcity['month'] = concatcity['month'].replace(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
 ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August','September', 'October', 'November','December'])
concatcity

month  year       city_name  order_item
0     October  2016  rio de janeiro          10
1     October  2016       sao paulo          31
0     January  2017  rio de janeiro          13
1     January  2017       sao paulo          65
2    February  2017           araxa           2
..        ...   ...             ...         ...
21       July  2018       sao paulo         598
22     August  2018       fronteira           3
23     August  2018  rio de janeiro          27
24     August  2018       sao paulo         779
25  September  2018       sao paulo          13

[71 rows x 4 columns]

Agar memudahkan proses visualisasi data frame yang terpecah disatukan menjadi satu dataframe denga menggunakan perintah concate(). untuk kolom month akan diubah menjadi bentuk nama bulan. karena kolom month masih bertipe data integer sedangkan untuk di convert ke nama bulan tipe data harus string, maka ubah dahulu tipe data menjadi string dengan perintah astype(str). kemudian convert data month menjadi nama bulan menggunakan perintah replace()

##**Persebaran seller berdasarkan kota**

In [ ]:
data = pd.DataFrame(data[['seller_id','city_name']])
data

seller_id       city_name
shipping_limit_date                                                  
2017-12-04 07:16:52  dd55f1bb788714a40e7954c3be6df745       sao paulo
2017-12-01 20:52:12  dd55f1bb788714a40e7954c3be6df745       sao paulo
2018-01-19 20:09:18  dd55f1bb788714a40e7954c3be6df745       sao paulo
2018-05-14 03:52:03  dd55f1bb788714a40e7954c3be6df745       sao paulo
2017-12-20 13:11:54  dd55f1bb788714a40e7954c3be6df745       sao paulo
...                                               ...             ...
2017-01-21 16:59:30  46dc3b2cc0980fb8ec44634e21d2718e  rio de janeiro
2018-01-29 13:01:14  46dc3b2cc0980fb8ec44634e21d2718e  rio de janeiro
2017-11-23 12:26:22  46dc3b2cc0980fb8ec44634e21d2718e  rio de janeiro
2017-12-13 22:54:28  46dc3b2cc0980fb8ec44634e21d2718e  rio de janeiro
2017-09-25 03:33:31  46dc3b2cc0980fb8ec44634e21d2718e  rio de janeiro

[10169 rows x 2 columns]

In [ ]:
data = data.groupby('city_name')['seller_id'].aggregate(['count']).reset_index()
data['total_seller'] = data['count']
data = data.drop('count',axis=1)
data

city_name  total_seller
0           araxa             4
1       fronteira            46
2   itau de minas             2
3          passos             1
4  rio de janeiro           622
5       sao paulo          9446
6         uberaba            48

# Visualisasi

##Visualisasi jumlah order item berdasarkan seller

In [ ]:
import plotly.express as px
df = concatsel
fig = px.treemap(df, path=[px.Constant("order item by seller"), 'year','month', 'seller_id', 'order_item'], values='order_item', color='seller_id')
fig.update_traces(marker=dict(cornerradius=5))
fig.show()

visualisasi data menggunakan treemap chart agar distribusi data mudah untuk dibaca. berdasarkan data tahun 2016 seller yang memiliki penjualan terbanyak adalah dengan seller_id 46dc3b2cc0980fb8ec44634e21d2718e sebanyak 10 order_item pada bulan oktober. pada tahun 2017 seller yang memiliki penjualan tertinggi dengan seller_id 46dc3b2cc0980fb8ec44634e21d2718e dengan jumlah 106 order_item pada bulan desember. pada tahun 2018 seller yang memiliki penjualan paling tinggi adalah dengan seller_id 1025f02d44d7041d6cm8b6550bfa dengan jumlah penjualan 140 order_item pada bulan mei

##Visualisasi jumlah order item berdasarkan kota

In [ ]:
import plotly.express as px
df = concatcity
fig = px.treemap(df, path=[px.Constant("jumlah order item berdasarkan kota"), 'year','month', 'city_name', 'order_item'], values='order_item', color='city_name')
fig.update_traces(marker=dict(cornerradius=5))
fig.show()

visualisasi data menggunakan treemap chart agar distribusi data mudah untuk dibaca. berdasarkan data diatas pada tahun 2016,2017 dan 2018 kota sao paulo merupakan kota yang memiliki jumlah order_item terbanyak dengan jumlah 31 order_item pada bulan oktober 2016, 588 order_item pada desember 2017 dan 815 order_item pada mei 2018.

##Visualisasi persebaran seller berdasarkan Kota

In [ ]:
import plotly.express as px
df = data
fig = px.treemap(df, path=[px.Constant("seller by city"), 'city_name','total_seller'], values='total_seller', color='city_name')
fig.update_traces(marker=dict(cornerradius=5))
fig.show()

visualisasi data menggunakan treemap chart agar distribusi data mudah untuk dibaca. berdasarkan data diatas kota sao paulo merupakan kota yang memiliki persebaran seller terbanyak dengan jumlah 446 seller. urutan kedua ada pada kota rio de jenerio dengan jumlah 622 seller. urutan ke tiga  ada pada kota uberaba dengan jumlah 48 seller.